api/test/opentelemetry/trace/tracer_test.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'test_helper'
+
+describe OpenTelemetry::Trace::Tracer do
+  let(:invalid_span) { OpenTelemetry::Trace::Span::INVALID }
+  let(:tracer) { OpenTelemetry::Trace::Tracer.new }
+
+  describe '#current_span' do
+    let(:current_span) { tracer.start_span('current') }
+
+    it 'returns an invalid span by default' do
+      tracer.current_span.must_equal(invalid_span)
+    end
+
+    it 'returns the current span when set' do
+      OpenTelemetry::Context.with(
+        OpenTelemetry::Trace::Tracer.const_get(:CONTEXT_SPAN_KEY),
+        current_span
+      ) do
```

Comment:
I feel like this is testing at the wrong level of abstraction. The `OpenTelemetry::Context` is really an internal concern, along with the `Tracer::CONTEXT_SPAN_KEY`. It should be possible to test the semantics of `tracer.current_span` using only the public API of `Tracer`. In particular, the `in_span` convenience method, `with_span` and `current_span` are all related, and I think it makes sense to test them as a unit rather than using gymnastics to test them separately.

Reply:
@fbogsany it's a clunky test for sure. Let me poke at it a bit and see if I can make it better. Thanks for the guidance.

Reply:
@fbogsany I've settled on using `with_span` to test `current_span` so I moved the `current_span` expectation up to this `describe` block and removed it from the `#with_span` block. I think that makes it clearer what I'm testing.